### ISSUE TITLE
Tekton reconcile failing with invalid list of done tasks; some tasks were indicated completed without ancestors being done

### ISSUE THREAD

**Issue by NUEB3** [tektoncd/pipeline#5402]
# Expected Behavior

We are using tekton on large scale for CICD pipelines. It is expected to have some reconciles failures on tekton controller, due to any kind of erratic behavior. However, this failures should be a small percentage of the total number of reconciles.

# Actual Behavior

The majority of reconciles failures are due to some tasks were indicated completed without ancestors being done on the tekton controller. In our logs it represents almost 90% of the errors and it is equivalent to something close to 25% of the total reconciles results.

# Steps to Reproduce the Problem

There is no specific way to reproduce it.

# Additional Info

- Kubernetes version:

  **Output of `kubectl version`:**

  ```
  Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.4", GitCommit:"b695d79d4f967c403a96986f1750a35eb75e75f1", GitTreeState:"clean", BuildDate:"2021-11-17T15:48:33Z", GoVersion:"go1.16.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"22+", GitVersion:"v1.22.10-eks-84b4fe6", GitCommit:"cc6a1b4915a99f49f5510ef0667f94b9ca832a8a", GitTreeState:"clean", BuildDate:"2022-06-09T18:24:04Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
  ```

- Tekton Pipeline version:

  **v0.35.1**

- Error Trace:
``` json 
{"level":"error","ts":"2022-08-31T14:23:38.918Z","logger":"tekton-pipelines-controller","caller":"controller/controller.go:566","msg":"Reconcile error","commit":"422a468","knative.dev/controller":"github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.Reconciler","knative.dev/kind":"tekton.dev.PipelineRun","knative.dev/traceid":"38774d09-5828-4687-88db-631a7d18abab","knative.dev/key":"pipelines/itaipu-tests-main-run-8gpdx","duration":0.114080912,"error":"1 error occurred:\n\t* invalid list of done tasks; some tasks were indicated completed without ancestors being done: [unit-test-00 unit-test-01 unit-test-02 unit-test-03 unit-test-04 unit-test-05 unit-test-06 unit-test-07 unit-test-08 unit-test-09 unit-test-10 unit-test-11 unit-test-12 unit-test-13 unit-test-14 unit-test-15]\n\n","stacktrace":"github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller.(*Impl).handleErr\n\tgithub.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:566\ngithub.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller.(*Impl).processNextWorkItem\n\tgithub.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:543\ngithub.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller.(*Impl).RunContext.func3\n\tgithub.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:491"}
```

**Comment by lbernick** [comment]
This looks very similar to https://github.com/tektoncd/pipeline/issues/4840 although that issue was fixed in v0.35.1-- can you double check that's the version you're using? Are you able to provide the PipelineRun spec for your testing Pipeline? This will be challenging to fix without an example PipelineRun yaml.

Side note, you might want to check out the [matrix](https://tekton.dev/docs/pipelines/matrix/#fan-out) feature just based on what your pipeline appears to be doing in this log line.

**Comment by pritidesai** [comment]
Thank you @NUEB3 for reporting this issue!

The error message has listed tasks `unit-test-00` through `unit-test-15`. Please provide an additional details around dependencies:

- Does all of these tasks depend on a single parent or a common set of tasks?
- Is the dependency defined through `runAfter`, task results, etc?
- Are all `unit-test-*` running in parallel? 
- Are there multiple runs happening with the same pipeline?

Any additional details on the pipeline definition will help troubleshoot this issue. Thanks!
